File: backend/app/models/recipe.py
"""Pydantic models for recipe API."""

from pydantic import BaseModel, HttpUrl, field_validator


class Ingredient(BaseModel):
    amount: str = ""
    item: str

    @field_validator("amount", mode="before")
    @classmethod
    def convert_none_to_empty(cls, v):
        """Convert None to empty string for ingredients without amounts."""
        return "" if v is None else str(v)


class Instruction(BaseModel):
    step: int
    text: str


class ConvertRequest(BaseModel):
    url: HttpUrl


class RecipeResponse(BaseModel):
    id: str
    youtubeUrl: str
    videoId: str
    videoTitle: str | None = None
    channelName: str | None = None
    creatorUsername: str | None = None
    creatorUrl: str | None = None
    platform: str | None = None  # "youtube" or "tiktok"
    title: str
    ingredients: list[Ingredient]
    instructions: list[Instruction]
    prepTime: str | None = None
    cookTime: str | None = None
    servings: str | None = None
    tags: list[str] = []  # AI-suggested tags
    cached: bool = False


class ErrorResponse(BaseModel):
    error: str
    detail: str | None = None

File: frontend/src/types/recipe.ts
export interface Ingredient {
  amount: string;
  item: string;
}

export interface Instruction {
  step: number;
  text: string;
}

export interface Recipe {
  id: string;
  youtubeUrl: string;
  videoId: string;
  videoTitle: string | null;
  channelName: string | null;
  creatorUsername: string | null;
  creatorUrl: string | null;
  platform: "youtube" | "tiktok" | null;
  title: string;
  ingredients: Ingredient[];
  instructions: Instruction[];
  prepTime: string | null;
  cookTime: string | null;
  servings: string | null;
  tags?: string[];  // AI-suggested tags from backend
  cached: boolean;
}

// Available recipe tags/categories (must match backend AVAILABLE_TAGS)
export const RECIPE_TAGS = [
  // Dish type
  "entrada",
  "plato-fuerte",
  "postre",
  "salsa",
  "bebida",
  "desayuno",
  "snack",
  "sopa",
  "ensalada",
  "guarnicion",
  "pan",
  "masa",
  // Cuisine
  "mexicana",
  "italiana",
  "asiatica",
  "americana",
  "francesa",
  "mediterranea",
  // Diet
  "saludable",
  "rapida",
  "vegetariana",
  "vegana",
  "sin-gluten",
  "keto",
  // Flavor profile
  "picante",
  "dulce",
  "agridulce",
  "ahumado",
  // Protein
  "pollo",
  "carne",
  "cerdo",
  "mariscos",
  "pescado",
  // Cooking method
  "horno",
  "parrilla",
  "sarten",
  "olla",
  "freidora",
  "sin-coccion",
] as const;

export type RecipeTag = (typeof RECIPE_TAGS)[number];

// User's personal note on a recipe
export interface RecipeNote {
  id: string;
  text: string;
  createdAt: string;
}

// Extended recipe with user-specific data (stored in localStorage)
export interface SavedRecipe extends Recipe {
  tags: RecipeTag[];
  notes: RecipeNote[];
  collectionIds?: string[];
  // User edits override original values
  userEdits?: {
    title?: string;
    ingredients?: Ingredient[];
    instructions?: Instruction[];
    prepTime?: string | null;
    cookTime?: string | null;
    servings?: string | null;
  };
  savedAt: string;
}

export interface ConvertRequest {
  url: string;
}

export interface ErrorResponse {
  detail: string;
}
